// @angular/cli/commands/build.js
"use strict";
const version_1 = require('../upgrade/version');
const Command = require('../ember-cli/lib/models/command');
// defaults for BuildOptions
exports.baseBuildCommandOptions = [
    {
        name: 'target',
        type: String,
        default: 'development',
        aliases: ['t', { 'dev': 'development' }, { 'prod': 'production' }]
    },
    { name: 'environment', type: String, aliases: ['e'] },
    { name: 'output-path', type: 'Path', aliases: ['op'] },
    { name: 'aot', type: Boolean },
    { name: 'sourcemap', type: Boolean, aliases: ['sm', 'sourcemaps'] },
    { name: 'vendor-chunk', type: Boolean, default: true, aliases: ['vc'] },
    { name: 'base-href', type: String, aliases: ['bh'] },
    { name: 'deploy-url', type: String, aliases: ['d'] },
    { name: 'verbose', type: Boolean, default: false, aliases: ['v'] },
    { name: 'progress', type: Boolean, default: true, aliases: ['pr'] },
    { name: 'i18n-file', type: String },
    { name: 'i18n-format', type: String },
    { name: 'locale', type: String },
    { name: 'extract-css', type: Boolean, aliases: ['ec'] },
    {
        name: 'output-hashing',
        type: String,
        values: ['none', 'all', 'media', 'bundles'],
        description: 'define the output filename cache-busting hashing mode',
        aliases: ['oh']
    },
];
const BuildCommand = Command.extend({
    name: 'build',
    description: 'Builds your app and places it into the output path (dist/ by default).',
    aliases: ['b'],
    availableOptions: exports.baseBuildCommandOptions.concat([
        { name: 'watch', type: Boolean, default: false, aliases: ['w'] }
    ]),
    run: function (commandOptions) {
        const project = this.project;
        // Check angular version.
        version_1.Version.assertAngularVersionIs2_3_1OrHigher(project.root);
        const BuildTask = require('../tasks/build').default;
        const buildTask = new BuildTask({
            cliProject: project,
            ui: this.ui,
        });
        return buildTask.run(commandOptions);
    }
});
BuildCommand.overrideCore = true;
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = BuildCommand;
