// @angular/cli/commands/e2e.js
"use strict";
const SilentError = require('silent-error');
const config_1 = require('../models/config');
const serve_1 = require('./serve');
const Command = require('../ember-cli/lib/models/command');
exports.e2eCommandOptions = serve_1.baseServeCommandOptions.concat([
    { name: 'config', type: String, aliases: ['c'] },
    { name: 'specs', type: Array, default: [], aliases: ['sp'] },
    { name: 'element-explorer', type: Boolean, default: false, aliases: ['ee'] },
    { name: 'webdriver-update', type: Boolean, default: true, aliases: ['wu'] },
    { name: 'serve', type: Boolean, default: true, aliases: ['s'] }
]);
const E2eCommand = Command.extend({
    name: 'e2e',
    aliases: ['e'],
    description: 'Run e2e tests in existing project',
    works: 'insideProject',
    availableOptions: exports.e2eCommandOptions,
    run: function (commandOptions) {
        const E2eTask = require('../tasks/e2e').E2eTask;
        this.project.ngConfig = this.project.ngConfig || config_1.CliConfig.fromProject();
        const e2eTask = new E2eTask({
            ui: this.ui,
            project: this.project
        });
        if (!commandOptions.config) {
            const e2eConfig = config_1.CliConfig.fromProject().config.e2e;
            if (!e2eConfig.protractor.config) {
                throw new SilentError('No protractor config found in angular-cli.json.');
            }
            commandOptions.config = e2eConfig.protractor.config;
        }
        if (commandOptions.serve) {
            const ServeTask = require('../tasks/serve').default;
            const serve = new ServeTask({
                ui: this.ui,
                project: this.project,
            });
            // Protractor will end the proccess, so we don't need to kill the dev server
            return serve.run(commandOptions, () => e2eTask.run(commandOptions));
        }
        else {
            return e2eTask.run(commandOptions);
        }
    }
});
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = E2eCommand;
